pkg/pillar/zedcloud/authen.go

```diff
@@ -431,21 +420,59 @@ func VerifySigningCertChain(ctx *ZedCloudContext, content []byte) ([]byte, error
 	}
 
 	log.Debugf("VerifySigningCertChain: key count %d\n", keyCnt)
-	if leafcert == nil {
-		err := fmt.Errorf("VerifySigningCertChain: fail to acquire leaf cert")
+	if signKeyCnt == 0 {
+		err := fmt.Errorf("VerifySigningCertChain: fail to acquire signing cert")
+		return nil, err
+	}
+
+	if err := VerifySignature(certByte, interm); err != nil {
+		errStr := fmt.Sprintf("%v", err)
+		log.Errorf("VerifySigningCertChain: " + errStr)
+		return nil, err
+	}
+
+	if err := UpdateServerCert(ctx, certByte); err != nil {
+		errStr := fmt.Sprintf("%v", err)
+		log.Errorf("VerifySigningCertChain: " + errStr)
 		return nil, err
 	}
 
+	log.Debugf("VerifySigningCertChain: success\n")
+	return certByte, nil
+}
+
+func VerifySignature(certByte []byte, interm *x509.CertPool) error {
+	block, _ := pem.Decode(certByte)
+	if block == nil {
+		errStr := fmt.Sprintf("VerifySignature: certificate block decode fail")
+		return errors.New(errStr)
+	}
+
+	leafcert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		errStr := fmt.Sprintf("VerifySignature: certificate parse fail, %v\n", err)
+		return errors.New(errStr)
+	}
+
+	// append to intermediate certificate
+	if interm != nil {
+		ok := interm.AppendCertsFromPEM(certByte)
+		if !ok {
+			errStr := fmt.Sprintf("VerifySignature: certificate append fail")
+			return errors.New(errStr)
+		}
+	}
+
```

Comment:
The certByte is the leaf certificate. You shouldn't add that to the intermediates.

Reply:
Sure. Done.